ios/RN/RNCamera.m

```diff
@@ -421,40 +403,110 @@ - (void)record:(NSDictionary *)options resolve:(RCTPromiseResolveBlock)resolve r
-        self.videoRecordedReject = reject;
     });
 }
 
-- (void)stopRecording
+- (void)stopAssetWriter {
+    self.canAppendBuffer = NO;
+    [self.writerInput markAsFinished];
+    [self.videoWriter finishWritingWithCompletionHandler:^{
+        if (self.videoWriter.status != AVAssetWriterStatusFailed) {
+            [self processVideoToAnimation:self.videoWriter.outputURL];
+        } else if (self.videoRecordedReject != nil) {
+            self.videoRecordedReject(@"E_RECORDING_FAILED", @"An error occurred while recording a video.", nil);
+        }
+
+        [self.session removeOutput:self.videoOutput];
+        self.videoOutput = nil;
+        self.writerInput = nil;
+        self.videoWriter = nil;
+    }];
+}
+
+- (void)captureOutput:(AVCaptureOutput *)output didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection
+{
+    [self findPrimaryFace:sampleBuffer];
+
+    if (self.canAppendBuffer) {
+        if (self.videoWriter.status != AVAssetWriterStatusWriting) {
+            [self.videoWriter startWriting];
+            [self.videoWriter startSessionAtSourceTime:CMSampleBufferGetPresentationTimeStamp(sampleBuffer)];
+            dispatch_async(dispatch_get_main_queue(), ^{
+                self.timer = [NSTimer scheduledTimerWithTimeInterval:self.maxDuration target:self selector:@selector(stopAssetWriter) userInfo:nil repeats:NO];
+            });
+        }
+        [self.writerInput appendSampleBuffer:sampleBuffer];
+    }
+}
+
+-(void)findPrimaryFace:(CMSampleBufferRef)sampleBuffer {
+    CVPixelBufferRef pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer);
+    CIImage *image = [CIImage imageWithCVPixelBuffer:pixelBuffer];
+    CIImage *orientedImage = [image imageByApplyingCGOrientation:kCGImagePropertyOrientationUpMirrored];
+
+    VNDetectFaceRectanglesRequest *faceDetectionReq = [VNDetectFaceRectanglesRequest new];
+    NSDictionary *d = [[NSDictionary alloc] init];
+    VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithCIImage:orientedImage options:d];
+    [handler performRequests:@[faceDetectionReq] error:nil];
+
+    VNFaceObservation *primaryFace;
+    CGPoint primaryFaceCenter;
+    float primaryFaceSize;
+
+    for(VNFaceObservation *observation in faceDetectionReq.results){
+        if(observation){
+            float size = observation.boundingBox.size.height * observation.boundingBox.size.width;
+            if (!primaryFace || size > primaryFaceSize) {
+                primaryFace = observation;
+                primaryFaceCenter = CGPointMake(CGRectGetMidX(observation.boundingBox), CGRectGetMidY(observation.boundingBox));
+                primaryFaceSize = size;
+            }
+        }
+    }
+
+    if ([faceDetectionReq.results count] && !self.exposureTimeout) {
+        [self setExposure:primaryFaceCenter];
+        self.exposureTimeout = YES;
+        dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
Is this being done to prevent trying to set the exposure while it's already in progress? If so, perhaps update the `setExposure` method to check whether the device is already locked and silently bail. Unless the NSTimer is serving another purpose?

Reply:
Moved this check into the `setExposure` method